### PR TITLE
[SDK/CLI] Support connection provider config

### PR DESCRIPTION
**Pull request by brynn-code**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.
pf config set & pf config show
image

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_cli/_pf/_config.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+from promptflow._cli._params import add_param_set_positional, logging_params
+from promptflow._cli._utils import activate_action, list_of_dict_to_dict
+from promptflow._sdk._configuration import Configuration
+from promptflow._sdk._constants import LOGGER_NAME
+
+logger = logging.getLogger(LOGGER_NAME)
+
+
+def add_config_set(subparsers):
```
Comment (brynn-code) [inline src/promptflow/promptflow/_cli/_pf/_config.py:13]: Mark: addtest

**Comment by brynn-code** [comment]
add changelog.

File: src/promptflow/promptflow/_cli/_pf/_config.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    # Config connection provider to azure workspace for current user:
+    pf config set connection.provider="azureml:/subscriptions/<your-subscription>/resourceGroups/<your-resourcegroup>/providers/Microsoft.MachineLearningServices/workspaces/<your-workspace>"
+    """  # noqa: E501
+    activate_action(
+        name="set",
+        description="Set promptflow configs for current user.",
+        epilog=epilog,
+        add_params=[add_param_set_positional] + logging_params,
+        subparsers=subparsers,
+        help_message="Set promptflow configs for current user.",
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_cli/_pf/_config.py:26]: include the config yaml path in help?

File: src/promptflow/promptflow/_cli/_pf/_connection.py [hunk 3]
```diff
@@ -29,7 +29,11 @@ def add_param_name(parser, required=False):
 
 def add_connection_parser(subparsers):
     connection_parser = subparsers.add_parser(
-        "connection", description="A CLI tool to manage connections for promptflow.", help="pf connection"
+        "connection",
+        description="""A CLI tool to manage connections for promptflow.
+
+        Your secrets will be encrypted using AES-128-CBC(https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) technology.""",  # noqa: E501
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_cli/_pf/_connection.py:35]: no need to have the link here? does it clickable in cli?

File: .github/workflows/promptflow-global-config-test.yml [hunk 4]
```diff
@@ -0,0 +1,113 @@
+name: promptflow-global-config-test
+on:
+  schedule:
+    - cron: "40 18 * * *" # Every day starting at 2:40 BJT
+  pull_request_target:
```
Comment (crazygao) [inline .github/workflows/promptflow-global-config-test.yml:5]: When test, use pull_request trigger to test your own pr.
When check in, please change back to pull_request_target.

File: src/promptflow/promptflow/_cli/_pf/_connection.py (hunk 3 above)
Comment (wangchao1230) [reply src/promptflow/promptflow/_cli/_pf/_connection.py:35]: but it's good to put in doc.

File: .github/workflows/promptflow-global-config-test.yml [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      'external' || 'internal' }}
+    runs-on: ubuntu-latest
+    steps:
+      - run: true
+  sdk_cli_tests:
+    needs: authorize
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, macos-latest]
```
Comment (crazygao) [inline .github/workflows/promptflow-global-config-test.yml:31]: Not necessary macos test

File: .github/workflows/promptflow-global-config-test.yml [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  authorize:
+    environment:
+      # forked prs from pull_request_target will be run in external environment, domain prs will be run in internal environment
+      ${{ github.event_name == 'pull_request_target' &&
+      github.event.pull_request.head.repo.full_name != github.repository &&
+      'external' || 'internal' }}
+    runs-on: ubuntu-latest
+    steps:
+      - run: true
+  sdk_cli_tests:
```
Comment (crazygao) [inline .github/workflows/promptflow-global-config-test.yml:26]: change to your own job name

File: .github/workflows/promptflow-global-config-test.yml [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+      if: always()
+      uses: actions/upload-artifact@v3
+      with:
+        name: Test Results (Python ${{ steps.display_env.outputs.pyVersion }}) (OS ${{ matrix.os }})
+        path: |
+          ${{ env.testWorkingDirectory }}/*.xml
+          ${{ env.testWorkingDirectory }}/htmlcov/
+  publish-test-results:
+    name: "Publish Tests Results"
+    needs: sdk_cli_tests
```
Comment (crazygao) [inline .github/workflows/promptflow-global-config-test.yml:94]: change accordingly

File: src/promptflow/promptflow/_sdk/operations/_local_azure_connection_operations.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        return self._pfazure_client._connections.get(name)
+
+    def delete(self, name: str) -> None:
+        """Delete a connection entity.
+
+        :param name: Name of the connection.
+        :type name: str
+        """
+        raise NotImplementedError(
+            "Delete connection is not supported for workspace connection, please go to Azure Portal to delete it."
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_local_azure_connection_operations.py:91]: Not sure if this is better.
Delete workspace connection is not supported in promptflow, please manage it in workspace portal, az ml cli or AzureML SDK?

File: src/promptflow/tests/sdk_cli_global_config_test/e2etests/test_global_config.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+DATAS_DIR = Path(__file__).parent.parent.parent / "test_configs" / "datas"
+
+
+@pytest.mark.usefixtures("global_config")
+@pytest.mark.e2etest
+class TestGlobalConfig:
+    def test_basic_flow_bulk_run(self, pf) -> None:
+        data_path = f"{DATAS_DIR}/webClassification3.jsonl"
+
+        pf.run(flow=f"{FLOWS_DIR}/web_classification", data=data_path)
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_global_config_test/e2etests/test_global_config.py:15]: why multiple run without assert?

Comment (brynn-code) [reply src/promptflow/tests/sdk_cli_global_config_test/e2etests/test_global_config.py:15]: I copied test from sdk_cli_test/test_flow_run.py, lol

File: src/promptflow/promptflow/_sdk/_pf_client.py [hunk 10]
```diff
@@ -161,6 +170,13 @@ def runs(self) -> RunOperations:
     @property
     def connections(self) -> ConnectionOperations:
         """Connection operations that can manage connections."""
+        if not self._connections:
+            if self._connection_provider == ConnectionProvider.LOCAL.value:
+                logger.debug("Using local connection operations.")
+                self._connections = ConnectionOperations()
+            else:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_pf_client.py:177]: if _connection_provider starts with "azure"
else: raise not support error?

File: src/promptflow/tests/sdk_cli_test/unittests/test_config.py [hunk 11]
```diff
@@ -37,3 +42,16 @@ def test_get_or_set_installation_id(self, config):
     def test_config_instance(self, config):
         new_config = Configuration.get_instance()
         assert new_config is config
+
+    def test_get_workspace_from_config(self):
+        # Test config within flow folder
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/unittests/test_config.py:47]: add some test on invalid connection_provider?

File: .github/workflows/promptflow-global-config-test.yml [hunk 12]
```diff
@@ -0,0 +1,115 @@
+name: promptflow-global-config-test
+on:
+  push:
+    branches:
+      - brynn/support-config-connection-provider
```
Comment (brynn-code) [inline .github/workflows/promptflow-global-config-test.yml:5]: remove this

**Review by wangchao1230 — APPROVED** [review]

**Review by wangchao1230 — DISMISSED** [review]

File: src/promptflow/promptflow/_sdk/_constants.py [hunk 13]
```diff
@@ -48,6 +48,10 @@
 LINE_NUMBER = "line_number"
 
 AZUREML_PF_RUN_PROPERTIES_LINEAGE = "azureml.promptflow.input_run_id"
+AZURE_WORKSPACE_REGEX_FORMAT = (
+    "^azureml:[/]{1,2}subscriptions/([^/]+)/resource(groups|Groups)/([^/]+)"
+    "/providers/Microsoft.MachineLearningServices/workspaces/([^/]+)$"
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_constants.py:53]: allow user not specify providers

**Review by crazygao — DISMISSED** [review]

**Review by wangchao1230 — DISMISSED** [review]

**Review by zhengfeiwang — APPROVED** [review]

**Review by D-W- — APPROVED** [review]

**Pull request merged**